src/duplicates_impl.rs

```diff
@@ -89,7 +89,7 @@ mod private {
             // far), plus (hi - pending) / 2 pairs of never seen before items.
             let hi = hi.map(|hi| {
                 let max_pending = std::cmp::min(self.meta.pending, hi);
-                let max_new = std::cmp::max(hi - self.meta.pending, 0) / 2;
+                let max_new = hi.saturating_sub(self.meta.pending) / 2;
                 max_pending + max_new
```

Comment:
Hi there, one thing that bothers me is that `cmp::min` and `saturating_sub` actually answer the same question twice. (Namely: is `hi <= self.meta.pending`?)

I think it would be clearer to do this:

```
if hi <= self.meta.pending {
    // fewer or equally many iter-remaining elements than pending elements
    // => at most, each iter-remaining element is matched
    hi
} else {
    // fewer pending elements than iter-remaining elements
    // => at most:
    //    * each pending element is matched
    //    * the other iter-remaining elements come in pairs
    self.meta.pending + (hi-self.meta.pending)/2
}

```

Reply:
updated as per your suggestion